wolfcrypt/src/sp_int.c

```diff
@@ -4394,10 +4394,20 @@ int sp_init(sp_int* a)
  */
 int sp_init_size(sp_int* a, int size)
 {
-    int err = sp_init(a);
+    int err = MP_OKAY;
+
+    if (a == NULL) {
+        err = MP_VAL;
+    }
+    if (err == MP_OKAY) {
+    #ifdef HAVE_WOLF_BIGINT
+        wc_bigint_init(&a->raw);
+    #endif
+        _sp_zero(a);
+    }
 
     if (err == MP_OKAY) {
-        a->size = size;
+        ((sp_int_minimal *)a)->size = size;
```

Comment:
Why is this needed?
I would have thought from the description that it would only need to be where dp is accessed.

Reply:
It's needed because the compiler will warn, `-Warray-bounds`, on any dereference of a pointer to type `foo` if `foo` was allocated for less than `sizeof(foo)`.  The warning isn't driven by whether the accessed element is out of bounds, but by whether the object as a whole was adequately allocated.
```
In function ‘_sp_zero’,
    inlined from ‘sp_init’ at /home/wolfbot/tmp/wolfssl_test_workdir.34488/wolfssl/wolfcrypt/src/sp_int.c:4380:9,
    inlined from ‘sp_init_size’ at /home/wolfbot/tmp/wolfssl_test_workdir.34488/wolfssl/wolfcrypt/src/sp_int.c:4397:15,
b330196c28 (<sean@wolfssl.com> 2021-01-27 11:42:44 +1000 16085)             sp_init_size(b, 1);
    inlined from ‘sp_prime_is_prime’ at /home/wolfbot/tmp/wolfssl_test_workdir.34488/wolfssl/wolfcrypt/src/sp_int.c:16085:13:
539ef512fc (<sean@wolfssl.com> 2021-02-08 14:19:25 +1000 4355)     a->used = 0;
/home/wolfbot/tmp/wolfssl_test_workdir.34488/wolfssl/wolfcrypt/src/sp_int.c:4355:13: error: array subscript ‘sp_int[0]’ is partly outside array bounds of ‘unsigned char[24]’ [-Werror=array-bounds]
 4355 |     a->used = 0;
      |     ~~~~~~~~^~~
[...]
```

It's good that the compiler warns for this code pattern -- it is usually a mistake when an object is underallocated.

Adding `struct sp_int_minimal` allows for warning-free (and non-bug-prone) manipulation of (deliberately) underallocated `sp_int`s.